Proposal: Remove develop branch

guarani opened the pull request:
This is a proposal to switch Gutenberg Mobile over from a Git Flow to a "GitHub Flow" branch workflow.

Instead of using both a `develop` and a `trunk` branch as we currently do, we could instead use the same branch model Gutenberg uses where a single `trunk` branch contains code that has passed review ([this Gutenberg doc roughly describes the flow](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/git-workflow.md)).

Gutenberg Mobile releases will still be specified by tagged commits on the `trunk` branch. 

### Why

We often have multiple Gutenberg Releases "in action". For example, at time of writing `1.47.2` (a hotfix) is in the current production version of the main apps (16.8) while `1.48.1` (a betafix) is in the current beta version of the main apps (16.9). If the need for another release arises we can't simply look at `trunk` to say "that's what we should base our new release off". So instead we consider whether this is a hotfix or a betafix and find the appropriate tag to create our new release off of. This means that in practice, we don't normally look at the `trunk` branch – it's tags we use to know what code is in what release.

### Upsides
- Simplified branching model
- The Gutenberg Mobile release branches would only need to be merged back to `trunk` (instead of `trunk` and `develop`)
- We'd move this project closer to Gutenberg by adopting its branching model
- One less branch to wrangle

### Downsides
- Release tags would still happen on `trunk` in Gutenberg Mobile, normal feature branches would also be merged there, too. So if keeping a separate branch only for releases is useful, this change could be a negative in that regard.

### Changes needed
- [x] Update release checklist template
- [ ] Update release scripts
- [ ] Update [CircleCI config](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/.circleci/config.yml)
- [ ] Change the Gutenberg Mobile default branch from `develop` to `trunk`

Commit:
Proposal: Remove develop branch

guarani commented:
Are there any other docs, scripts, processes, etc that would need to be updated?

fluiddot commented:
> Are there any other docs, scripts, processes, etc that would need to be updated?

I was thinking about reviewing the steps of the "Last Day" and "Bringing release changes back to development branches" regarding the Gutenberg-mobile PR.

If we want to follow the same flow we do in Gutenberg, we should create the GitHub release and tag before the release branch is merged into `trunk` so it would be:

```
<!-- wp:heading {"level":3} -->
<h3>Last Day</h3>
<!-- /wp:heading -->

<!-- wp:paragraph -->
<p>o Make sure that the bundle files on the Gutenberg-Mobile release branch have been updated to include any changes to the release branch.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>o Create a new GitHub release pointing to the tag: https://github.com/wordpress-mobile/gutenberg-mobile/releases/new?tag=vX.XX.X&target=release/X.XX.X&title=Release%20X.XX.X. Include a list of changes in the release's description</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>o Tag the head of Gutenberg release branch that the Gutenberg-Mobile release branch is pointing to with the <code>rnmobile/X.XX.X</code> tag.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>o In WPiOS update the reference to point to the <em>tag</em>.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>o In WPAndroid, update the submodule to point to the merge commit on GB-Mobile <code>trunk</code>.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>o Main apps PRs should be ready to merge to their develop now. Merge them or get them merged.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>o Once everything is merged, ping our friends in #platform9 and let them know we’ve merged our release so everything is right from our side to cut the main app releases.</p>
<!-- /wp:paragraph -->

<!-- wp:heading {"level":3} -->
<h3>Bringing release changes back to development branches</h3>
<!-- /wp:heading -->

<!-- wp:paragraph -->
<p>o If there are any conflicts in the Gutenberg-Mobile PR, merge <code>trunk</code> into it and resolve them.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>o If there are any conflicts in the Gutenberg PR, merge <code>trunk</code> into it and resolve them.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>o Merge the Gutenberg PR to <code>trunk</code> and Gutenberg-Mobile PR to <code>trunk</code>.</p>
<!-- /wp:paragraph -->

<!-- wp:heading {"level":3} -->
<h3>AFTER the main apps have cut their release branches</h3>
<!-- /wp:heading -->

...
```

guarani commented:
> If we want to follow the same flow we do in Gutenberg, we should create the GitHub release and tag before the release branch is merged into trunk so it would be:

Yes, good point! One I'd overlooked when making this PR draft. If we tagged after merging, we would be picking up untested changes in the release (changes which had been merged to `trunk` between the time of the release cut and the tagging of the release).

Commit:
Update which branch should be tagged

If we rename `develop` to `trunk` and remove the current `trunk` branch, we must tag releases on the release branch instead of `trunk` as we currently do.
Otherwise releases will pick up untested changes in `trunk` when they are tagged.

Commit:
Merge branch 'develop' into try/remove-develop-branch

Commit:
Revert change that is outside scope

This change will be addressed in https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/pull/35

mchowning commented on Releasing.md:
```diff
@@ -160,14 +156,6 @@ For the body of the post, just copy this checklist and again replace all occurre
 <p>o If there are any conflicts in the Gutenberg PR, merge <code>trunk</code> into it and resolve them.</p>
```
I think we need a step like this for Gutenberg-Mobile somewhere too since there could be conflicts between the release branch and `trunk` there as well.

mchowning replied:
Also, and this is not your change, but it looks like there are two `Merge the Gutenberg PR to <code>trunk</code>` steps (lines 160 and 168). I'm thinking the second one should be `Gutenberg-Mobile` instead of `Gutenberg`.

mchowning commented on Releasing.md:
```diff
@@ -160,14 +156,6 @@ For the body of the post, just copy this checklist and again replace all occurre
-<!-- /wp:paragraph -->
-
 <!-- wp:paragraph -->
 <p>o Merge the Gutenberg PR to <code>trunk</code>.</p>
```
This is not your change, but it looks like there are two `Merge the Gutenberg PR to <code>trunk</code>` steps (here and 8 lines below). I'm thinking the second one should be `Gutenberg-Mobile` instead of `Gutenberg`.

jhnstn commented:
Closing since we no longer use a develop branch.
